filament/backend/src/opengl/platforms/CocoaExternalImage.mm

```diff
@@ -0,0 +1,236 @@
+    glUniform1i(samplerLoc, 0);
+
+    // Restore state.
+    glUseProgram(currentProgram);
+}
+
+CocoaExternalImage::SharedGl::~SharedGl() noexcept {
+    glDeleteSamplers(1, &sampler);
+    glDetachShader(program, vertexShader);
+    glDetachShader(program, fragmentShader);
+    glDeleteShader(vertexShader);
+    glDeleteShader(fragmentShader);
+    glDeleteProgram(program);
+}
+
+CocoaExternalImage::CocoaExternalImage(const CVOpenGLTextureCacheRef textureCache,
+        const SharedGl &sharedGl) noexcept : mSharedGl(sharedGl), mTextureCache(textureCache) {
+    glGenFramebuffers(1, &mFBO);
+    CHECK_GL_ERROR(utils::slog.e)
+}
+
+CocoaExternalImage::~CocoaExternalImage() noexcept {
+    glDeleteFramebuffers(1, &mFBO);
+    release();
+}
+
+bool CocoaExternalImage::set(CVPixelBufferRef image) noexcept {
+    // Release references to a previous external image, if we're holding any.
+    release();
+
+    if (!image) {
+        return false;
+    }
+
+    OSType formatType = CVPixelBufferGetPixelFormatType(image);
+    ASSERT_POSTCONDITION(formatType == kCVPixelFormatType_32BGRA,
+            "macOS external images must be in either 32BGRA format.");
+
+    // The pixel buffer must be locked whenever we do rendering with it. We'll unlock it before
+    // releasing.
+    UTILS_UNUSED_IN_RELEASE CVReturn lockStatus = CVPixelBufferLockBaseAddress(image, 0);
+    assert_invariant(lockStatus == kCVReturnSuccess);
+
+    mImage = image;
+    mTexture = createTextureFromImage(image);
+    mRgbaTexture = encodeBlitTexture(CVOpenGLTextureGetName(mTexture),
+            CVPixelBufferGetWidth(image), CVPixelBufferGetHeight(image));
+    CHECK_GL_ERROR(utils::slog.e)
+
+    return true;
+}
+
+GLuint CocoaExternalImage::getGlTexture() const noexcept {
+    if (mEncodedToRgba) {
+        return mRgbaTexture;
+    }
+    return CVOpenGLTextureGetName(mTexture);
+}
+
+GLuint CocoaExternalImage::getInternalFormat() const noexcept {
+    return GL_RGBA8;
+}
+
+GLuint CocoaExternalImage::getTarget() const noexcept {
+    if (mEncodedToRgba) {
+        return GL_TEXTURE_2D;
+    }
+    return CVOpenGLTextureGetTarget(mTexture);
+}
+
+void CocoaExternalImage::release() noexcept {
+    if (mImage) {
+        CVPixelBufferUnlockBaseAddress(mImage, 0);
+        CVPixelBufferRelease(mImage);
+    }
+    if (mTexture) {
+        CFRelease(mTexture);
+    }
+    if (mEncodedToRgba) {
+        glDeleteTextures(1, &mRgbaTexture);
+        mRgbaTexture = 0;
+    }
+}
+
+CVOpenGLTextureRef CocoaExternalImage::createTextureFromImage(CVPixelBufferRef image) noexcept {
+    CVOpenGLTextureRef texture = nullptr;
+    UTILS_UNUSED_IN_RELEASE CVReturn success =
+            CVOpenGLTextureCacheCreateTextureFromImage(kCFAllocatorDefault,
+            mTextureCache, image, nil, &texture);
+    assert_invariant(success == kCVReturnSuccess);
+
+    return texture;
+}
+
+GLuint CocoaExternalImage::encodeBlitTexture(GLuint srcTexture, size_t width, size_t height) noexcept {
```

Comment:
if I understand correctly, this is used mainly to convert from GL_TEXTURE_RECTANGLE target to GL_TEXTURE_2D, right? it also does the swizzling from BGRA. If so, can you add a comment about that?

Reply:
Instead of adding comment I have changed a function name.